### PR TITLE
Notification consistency improvements

### PR DESCRIPTION
**Pull request by gmethvin**
This seems to fix some, but not all, of the problems we have with subsuming notifications. There are two main parts to this change:
- We now use the threadId on the extension to decide whether to subsume the notification. This deals with cases where we may have missed a "notification" message due to a disconnect. In this case we can get notifications through "missed_notifications", but those may reference subsumed notifications we don't have.
- Since we generate notifications asynchronously, the order in which the notifications are created can be different from the order of their corresponding messages. We should avoid sending notifications in this case, since they would be immediately subsumed anyway.

I think there are still a few possible race conditions here. I think if we can use "SELECT...FOR UPDATE" we can eliminate some of them. We could also consider using an actor or some other method to serialize creation of notifications.

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/realtime/UserNotifier.scala [hunk 1]
```diff
@@ -47,6 +47,7 @@ case class MessageDetails(
   createdAt: DateTime,
   newCount: Int,
   totalCount: Int,
+  threadId: String,
```
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/realtime/UserNotifier.scala]: Every time these `Details` classes change, we have to regenerate+persist the JSON for all past notifications. Andrew has written some code to do that, but I don't recall exactly where it is. It might be easier to simply parse the thread ID from the locator field. I know we're already doing that in at least one place in the extension. Or just compare the locator field directly, without bothering to parse out the thread ID. At some time in the past, I think we were using message IDs in locators, but now I'm pretty sure we're just using thread IDs.

File: kifi-backend/modules/shoebox/app/com/keepit/realtime/UserNotifier.scala [hunk 2]
```diff
@@ -318,33 +319,37 @@ class UserNotifier @Inject() (
...
-      val lastNotice = lastNotifiedMessage flatMap { lastMsg =>
-        userNotifyRepo.getWithCommentId(userId, lastMsg.id.get) map { oldNotice =>
-          userNotifyRepo.save(oldNotice.withState(UserNotificationStates.SUBSUMED))
-        }
+    val generatedSet = (for (userId <- participants - author.id.get) yield {
+      val lastMessage +: olderMessages = thread.filter(_.userId != userId)
+      val subsumed = olderMessages flatMap { m =>
+        userNotifyRepo.getWithCommentId(userId, m.id.get)
+      } map { oldNotice =>
+        userNotifyRepo.save(oldNotice.withState(UserNotificationStates.SUBSUMED))
```
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/realtime/UserNotifier.scala:324]: We expect many of the notifications loaded here to be `SUBSUMED` already. Probably a good idea to filter them out before re-persisting, to reduce the number of writes, but also to avoid changing a row's `updatedAt` value when nothing has really changed. Since filtering is best done at the database layer, may want to pass `SUBSUMED` as an exclude state to `userNotifyRepo.getWithCommentId`.

Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/realtime/UserNotifier.scala:324]: Oh, I just looked it up and saw that `SUBSUMED` is already an exclude state by default for `getWithCommentId`. That was a surprise to me. I guess it might help the readability/understandability of call sites to pass it explicitly or use a method name that hints that subsumed notifications are excluded.

File: kifi-backend/modules/shoebox/app/com/keepit/realtime/UserNotifier.scala [hunk 3]
```diff
@@ -318,33 +319,37 @@ class UserNotifier @Inject() (
...
       }
+      if (lastMessage.id.get == message.id.get) {
+        val lastNotice = subsumed.find(_.id != message.id)
```
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/realtime/UserNotifier.scala]: This seems to be comparing an `Id[UserNotification]` with an `Id[Message]`. What am I missing?

Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/realtime/UserNotifier.scala]: Also, I think we're already guaranteed at this point that nothing in `subsumed` corresponds to `message`, since `message` is `lastMessage` and `subsumed` all originate from `olderMessages`.

File: kifi-backend/modules/shoebox/app/com/keepit/realtime/UserNotifier.scala [hunk 4]
```diff
@@ -318,33 +319,37 @@ class UserNotifier @Inject() (
...
-      (recipient, messageDetail, userNotifyRepo.save(UserNotification(
-        userId = userId,
-        category = UserNotificationCategories.MESSAGE,
-        details = UserNotificationDetails(Json.toJson(messageDetail)),
-        commentId = message.id,
-        subsumedId = lastNotice.map(_.id.get))))
-    }
+        val recipient = userRepo.get(userId)
+        val deepLink = deepLinkRepo.save(DeepLink(
+          initatorUserId = Option(message.userId),
```
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/realtime/UserNotifier.scala]: `initator` → `initiator` (an old spelling error)

File: packrat/main.js [hunk 5]
```diff
@@ -651,6 +651,9 @@ function insertNewNotification(n) {
         return false;
       }
       break;
+    } else if (n.details.threadId && sameThread(notifications[i], n)) {
```
Comment (2is10) [inline packrat/main.js]: This will generalize to comments as well if we use `.locator` instead of `.threadId`.

File: packrat/main.js [hunk 6]
```diff
@@ -664,19 +667,21 @@ function insertNewNotification(n) {
...
-  if (n.details.subsumes) {
-    for (i++; i < notifications.length; i++) {
-      var n2 = notifications[i];
-      if (n2.id == n.details.subsumes) {
-        notifications.splice(i, 1);
-        if (notificationNotVisited.test(n2.state)) {
-          decrementNumNotificationsNotVisited(n2);
-        }
-        break;
+  for (i++; i < notifications.length; i++) {
```
Comment (2is10) [inline packrat/main.js]: We could shorten this to `while (++i < notifications.length) {...}`.

File: packrat/main.js [hunk 7]
```diff
@@ -664,19 +667,21 @@ function insertNewNotification(n) {
     }
   }
 
-  if (n.details.subsumes) {
```
Comment (2is10) [inline packrat/main.js:667]: I'm not sure I understand the benefit of removing this condition. Under what circumstances could a notification that doesn't claim to subsume anything subsume something?

**Comment by 2is10** [comment]
Great detective work!

File: packrat/main.js (hunk 5 above)
Comment (gmethvin) [inline packrat/main.js]: good point

File: kifi-backend/modules/shoebox/app/com/keepit/realtime/UserNotifier.scala (hunk 3 above)
Comment (gmethvin) [inline kifi-backend/modules/shoebox/app/com/keepit/realtime/UserNotifier.scala]: Yeah, I'm not sure what I was thinking here. `subsumed.headOption` should be good.

File: packrat/main.js (hunk 7 above)
Comment (gmethvin) [inline packrat/main.js:667]: I think it's still possible to have a race condition in which we have two notifications for the thread, and the notification for the older message doesn't exist yet. In this case the subsumes property would be null.

Comment (2is10) [inline packrat/main.js:667]: Good point. Thanks.
